File: Python_language_practice/python_sorting_and_searching_techniques/quick_sort.py
def partition(array, low, high):
    k = (low - 1)
    pivot = array[high]
    for j in range(low, high):
        if array[j] < pivot:
            k = k + 1
            array[k], array[j] = array[j], array[k]
    array[k + 1], array[high] = array[high], array[k + 1]
    return k + 1


def quick_sort(array_list, low, high):
    if low < high:
        pi = partition(array_list, low, high)
        quick_sort(array_list, low, pi - 1)
        quick_sort(array_list, pi + 1, high)


if __name__ == "__main__":
    arr = list(map(int, input("enter elements: ").split()))
    n = len(arr)
    quick_sort(arr, 0, n - 1)
    print(arr)

File: datastructure/quick_sort.c
#include<stdio.h>
#include<stdbool.h>
#define MAX 10

int arr[MAX];
int size = sizeof(arr)/sizeof(arr[0]);

void swap(int* a, int* b)
{
    int t = *a;
    *a = *b;
    *b = t;
}

int partition(int low, int high)
{
    int pivot = arr[high];
    int i = low - 1;
    int j;
    for (j = low; j <= high; j++)
    {
        if (arr[j] < pivot)
        {
            i++;
            swap(&arr[i], &arr[j]);
        }
    }
    swap(&arr[i+1], &arr[high]);
    return(i+1);
}

void quicksort(int low, int high)
{
    if (low < high)
    {
        int temp = partition(low, high);
        quicksort(low, temp-1);
        quicksort(temp + 1, high);
    }

}

void display()
{
	int i;
    for (i = 0; i < size; i++)
    {
        printf("%d ",arr[i]);
    }
}

void insert()
{
	int i;
    for (i = 0; i < size; i++)
    {
        printf("enter %d element: ",i+1);
        scanf("%d",&arr[i]);
    }
}

int main()
{
    int option;
    bool flag = true;
    while (flag)
    {
        printf("enter 1 to insert array\n2 to quick sort\n3 to display\n4 to exit: ");
        scanf("%d",&option);
        switch (option)
        {
        case 1:
            insert();
            break;
        case 2:
            quicksort(0, size-1);
            break;
        case 3:
            display();
            break;
        case 4:
            flag = false;
            break;
        default:
            break;
        }
    }
    return 0;
}
